server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockSnapshotIndexInput.java

```diff
@@ -139,20 +141,30 @@ protected IndexInput fetchBlock(int blockId) throws IOException {
         // If the snapshot file is chunked, we must account for this by
         // choosing the appropriate file part and updating the position
         // accordingly.
-        final int part = (int) (blockStart / partSize);
-        final long partStart = part * partSize;
-
-        final long position = blockStart - partStart;
-        final long length = blockEnd - blockStart;
-
-        BlobFetchRequest blobFetchRequest = BlobFetchRequest.builder()
-            .position(position)
-            .length(length)
-            .blobName(fileInfo.partName(part))
-            .directory(directory)
-            .fileName(blockFileName)
-            .build();
-        return transferManager.fetchBlob(blobFetchRequest);
+        int partNum = (int) (blockStart / partSize);
+        long pos = blockStart;
+        long diff = (blockEnd - blockStart);
+
+        // Block may be present on multiple chunks of a file, so we need
+        // to make multiple blobFetchRequest to fetch an entire block.
+        // Each fetchBlobRequest can fetch only a single chunk of a file.
+        List<BlobFetchRequest.BlobPart> blobParts = new ArrayList<>();
+        while (diff > 0) {
```

Comment:
Can we pull this out into a package private method and test different scenarios with UTs?

Reply:
Sure I will pull this into a separate method and add some tests

Reply:
Addressed in latest commit